crates/providers/src/new.rs

```diff
@@ -138,8 +167,288 @@ pub trait Provider<N: Network, T: Transport + Clone = BoxTransport>: Send + Sync
+            },
+
+            Ok(None) => return Err(TransportErrorKind::custom_str("Latest block not found")),
+
+            Err(err) => return Err(err),
+        };
+
+        let fee_history = match self
+            .get_fee_history(
+                U256::from(utils::EIP1559_FEE_ESTIMATION_PAST_BLOCKS),
+                BlockNumberOrTag::Latest,
+                &[utils::EIP1559_FEE_ESTIMATION_REWARD_PERCENTILE],
+            )
+            .await
+        {
+            Ok(fee_history) => fee_history,
+            Err(err) => return Err(err),
+        };
+
+        // use the provided fee estimator function, or fallback to the default implementation.
+        let (max_fee_per_gas, max_priority_fee_per_gas) = if let Some(es) = estimator {
+            es(base_fee_per_gas, fee_history.reward.unwrap_or_default())
+        } else {
+            utils::eip1559_default_estimator(
+                base_fee_per_gas,
+                fee_history.reward.unwrap_or_default(),
+            )
+        };
+
+        Ok((max_fee_per_gas, max_priority_fee_per_gas))
+    }
+
+    async fn get_proof(
+        &self,
+        address: Address,
+        keys: Vec<StorageKey>,
+        block: Option<BlockId>,
+    ) -> TransportResult<EIP1186AccountProofResponse> {
+        self.client().prepare("eth_getProof", (address, keys, block.unwrap_or_default())).await
+    }
+
+    async fn create_access_list(
+        &self,
+        request: &N::TransactionRequest,
+        block: Option<BlockId>,
+    ) -> TransportResult<AccessListWithGasUsed> {
+        self.client().prepare("eth_createAccessList", (request, block.unwrap_or_default())).await
+    }
+
+    // todo: move to extension trait
+    /// Parity trace transaction.
+    async fn trace_transaction(
+        &self,
+        hash: TxHash,
+    ) -> TransportResult<Vec<LocalizedTransactionTrace>> {
+        self.client().prepare("trace_transaction", (hash,)).await
+    }
+
+    // todo: move to extension trait
+    async fn debug_trace_transaction(
+        &self,
+        hash: TxHash,
+        trace_options: GethDebugTracingOptions,
+    ) -> TransportResult<GethTrace> {
+        self.client().prepare("debug_traceTransaction", (hash, trace_options)).await
+    }
+
+    // todo: move to extension trait
+    async fn trace_block(
+        &self,
+        block: BlockNumberOrTag,
+    ) -> TransportResult<Vec<LocalizedTransactionTrace>> {
+        self.client().prepare("trace_block", (block,)).await
+    }
+}
+
+/// Extension trait for Anvil specific JSON-RPC methods.
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+pub trait AnvilProvider<N: Network, T: Transport + Clone = BoxTransport>: Provider<N, T> {
+    /// Set the bytecode of a given account.
+    async fn set_code(&self, address: Address, code: &'static str) -> TransportResult<()> {
+        self.client().prepare("anvil_setCode", (address, code)).await
+    }
```

Comment:
todo add other anvil methods and prob could also go to other crate in the future, for set_storage, mine etc.

Reply:
follow up

Reply:
https://github.com/alloy-rs/alloy/issues/271